onmt/translate/translator.py

```diff
@@ -1228,3 +1229,164 @@ def _score_target(
+            tokens = []
+            for t in range(batch_side.shape[0]):
+                token = field.vocab.itos[batch_side[t, i, 0]]
+                if token == field.pad_token or token == field.eos_token:
+                    break
+                if token != field.init_token:
+                    tokens.append(token)
+            tokenized_sentences.append(tokens)
+        return tokenized_sentences
+
+    def build_sources_and_refs(self, batch, mode):
+        """Reconstruct the sources and references of the examples
+        related to a batch"""
+        if mode == 'valid':
+            sources = []
+            refs = []
+            for example in batch.dataset.examples:
+                sources.append(example.src[0])
+                refs.append(example.tgt[0])
+        elif mode == 'train':
+            sources = self.tokenize_batch(batch.src[0], 'src')
+            refs = self.tokenize_batch(batch.tgt, 'tgt')
+        return sources, refs
+
+    def translate(self, model, batch, gpu_rank, step, mode):
+        """Compute the sentences predicted by the current model's state
+        related to a batch"""
+        model_opt = self.opt
+        parser = ArgumentParser()
+        onmt.opts.translate_opts(parser)
+        base_args = (["-model", "dummy"] + ["-src", "dummy"])
+        opt = parser.parse_args(base_args)
+        opt.gpu = gpu_rank
+        ArgumentParser.validate_translate_opts(opt)
+        ArgumentParser.update_model_opts(model_opt)
+        ArgumentParser.validate_model_opts(model_opt)
+        scorer = onmt.translate.GNMTGlobalScorer.from_opt(opt)
+
+        out_file = codecs.open(os.devnull, "w", "utf-8")
+        translator = Translator.from_opt(
+            model,
+            self.fields,
+            opt,
+            model_opt,
+            global_scorer=scorer,
+            out_file=out_file,
+            report_align=opt.report_align,
+            report_score=True,
+            logger=None)
+        sources, refs = self.build_sources_and_refs(batch, mode)
+        _, preds = translator.translate(
+            sources,
+            batch_size=model_opt.valid_batch_size,
+            batch_type=model_opt.batch_type)
+        texts_ref = []
+
+        for i in range(len(preds)):
+            preds[i] = self.tgt_detokenizer._detokenize(preds[i][0].split())
+            texts_ref.append(self.tgt_detokenizer._detokenize(refs[i]))
+
+        if len(preds) > 0 and self.opt.scoring_debug:
+            path = os.path.join(self.opt.dump_preds,
+                                "preds.{}_step_{}.{}".format(
+                                    mode, step, "txt"))
+            with open(path, "a") as file:
+                for i in range(len(preds)):
+                    file.write("SOURCE: {}\n".format(sources[i]))
+                    file.write("REF: {}\n".format(texts_ref[i]))
+                    file.write("PRED: {}\n\n".format(preds[i]))
+        return preds, texts_ref
```

Comment:
Maybe put this in some onmt.translate.utils (or better name) to prevent overcrowding the already crowded onmt.translate.translator